### ISSUE TITLE
[core] Support sharing properties across mutliple rules

### ISSUE THREAD

**Issue by adangel** [pmd/pmd#4331]
At the moment I need to specify duplicate violationSuppressXPath properties for multiple rules.

Ideally I would like to declare the properties once in a rule set then reference the properties from each rule where I would like to apply them (like path references in ant build.xml).

Otherwise another option would be to support properties being applied for all rules within a xml set.
e.g.

```xml
<rule ref="rulesets/java/codesize.xml">
    <properties>
      <property name="violationSuppressXPath" 
                value="//MethodDeclarator[@Image='equals'] | //MethodDeclarator[@Image='hashCode']"/>
    </properties>
  </rule>
```

instead of having to declare the properties for every rule within codesize.xml

_originally from https://sourceforge.net/p/pmd/bugs/1466/_

**Comment by uhafner** [comment]
Is this XPath value still valid for 7.x? I tried that for `CognitiveComplexity` and `NPathComplexity` without success. 

For `SimplifyBooleanReturns` I needed to use the following instead:
```
<rule ref="category/java/design.xml/SimplifyBooleanReturns">
    <properties>
      <property name="violationSuppressXPath" value="./ancestor::*[@Name = 'equals'] | ./ancestor::*[@Name = 'hashCode']"/>
    </properties>
  </rule>
```

**Comment by jsotuyod** [comment]
@uhafner no, in PMD 7 the `MethodDeclarator` node was removed, and name is available directly in the `MethodDeclaration` node. Using your solution, or a more precise `./ancestor::MethodDeclaration[@Name = 'equals'] | ./ancestor::MethodDeclaration[@Name = 'hashCode']` should work.

**Comment by uhafner** [comment]
Thanks, `./ancestor::MethodDeclaration[@Name = 'equals']` works as well, at least for `SimplifyBooleanReturns`. The metrics do not work in the same way, is this a bug?

```
  <rule ref="category/java/design.xml/CognitiveComplexity">
    <properties>
      <property name="violationSuppressXPath" value="./ancestor::MethodDeclaration[@Name = 'equals'] | ./ancestor::MethodDeclaration[@Name = 'hashCode']"/>
    </properties>
  </rule>
  <rule ref="category/java/design.xml/CyclomaticComplexity">
    <properties>
      <property name="violationSuppressXPath" value="./ancestor::MethodDeclaration[@Name = 'equals'] | ./ancestor::MethodDeclaration[@Name = 'hashCode']"/>
    </properties>
  </rule>
  <rule ref="category/java/design.xml/NPathComplexity">
    <properties>
      <property name="violationSuppressXPath" value="./ancestor::MethodDeclaration[@Name = 'equals'] | ./ancestor::MethodDeclaration[@Name = 'hashCode']"/>
    </properties>
  </rule>
```

**Comment by oowekyala** [comment]
Yeah that's because the metrics rules report on the MethodDeclaration node. So you have to use `.[@Name = 'equals']` or `./ancestor-or-self::MethodDeclaration[...]` if you want something that works on different rules.
